sdks/go/pkg/beam/io/spannerio/spanner.go

```diff
@@ -0,0 +1,163 @@
+package spannerio
+
+import (
+	"cloud.google.com/go/spanner"
+	"context"
+	"fmt"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam"
+	"google.golang.org/api/iterator"
+	"reflect"
+	"strings"
+)
+
+func init() {
+	beam.RegisterType(reflect.TypeOf((*queryFn)(nil)).Elem())
+	beam.RegisterType(reflect.TypeOf((*writeFn)(nil)).Elem())
+}
+
+func columnsFromStruct(t reflect.Type) []string {
+	var columns []string
+
+	for i := 0; i < t.NumField(); i++ {
+		columns = append(columns, t.Field(i).Tag.Get("spanner"))
+	}
+
+	return columns
+}
+
+// Read reads all rows from the given table. The table must have a schema
+// compatible with the given type, t, and Read returns a PCollection<t>. If the
+// table has more rows than t, then Read is implicitly a projection.
+func Read(s beam.Scope, database, table string, t reflect.Type) beam.PCollection {
+	s = s.Scope("spanner.Read")
+
+	// TODO(herohde) 7/13/2017: using * is probably too inefficient. We could infer
+	// a focused query from the type.
+
+	cols := strings.Join(columnsFromStruct(t), ",")
+
+	return query(s, database, fmt.Sprintf("SELECT %v from [%v]", cols, table), t)
+}
+
+// QueryOptions represents additional options for executing a query.
+type QueryOptions struct {
+}
+
+// Query executes a query. The output must have a schema compatible with the given
+// type, t. It returns a PCollection<t>.
```

Comment:
Since this is a naive implementation it would be good to document that the query will be processed on a single worker, to set performance expectations of the query, and to recommend that a beam.Reshuffle be added after this if downstream splitting is required. If this changes to an SDF, we can remove the note.

Reply:
Done